### PR TITLE
Add ESMValTool example metric

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

Add an example ESMValTool metric. First step towards #32.

## Checklist

Please confirm that this pull request has done the following:

- [ ] Tests added
- [ ] Documentation added (where applicable)
- [x] Changelog item added to `changelog/`

### CODE REVIEW

**Comment by bouweandela** [comment]
@lewisjared Is this the right place to start the `ref-metrics-esmvaltool` package? Or shall I put it in a separate repository under the [ESMValGroup](https://github.com/esMValGroup/) organization?

The tests could better, as could the functionality for editing recipes, running esmvaltool, and the CMEC output bundle, but maybe it's enough at this early stage.

Would it make sense to enable building the documentation for pull requests on readthedocs so they are easier to look at? Or are they available from the GitHub Actions and I just missed them?

File: packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/example.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+from typing import Any
+
+import xarray
+from ref_core.datasets import FacetFilter, SourceDatasetType
+from ref_core.metrics import DataRequirement, Metric, MetricExecutionDefinition, MetricResult
+from ruamel.yaml import YAML
```
Comment (lewisjared) [inline packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/example.py:6]: This is an aside, but why use ruamel over pyyaml?

File: packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/recipe.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
+from __future__ import annotations
+
+import subprocess
+from pathlib import Path
+from typing import TYPE_CHECKING, Any
+
+import pkg_resources
```
Comment (lewisjared) [inline packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/recipe.py:7]: pkg_resources is deprecated. Is it possible to replace this usage with `importlib.resources`?

File: packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/recipe.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        "ensemble": "member_id",
+        "exp": "experiment_id",
+        "grid": "grid_label",
+        "mip": "table_id",
+        "short_name": "variable_id",
+    },
+}
+
+
+def as_isodate(timestamp: pd._libs.tslibs.timestamps.Timestamp) -> str:
```
Comment (lewisjared) [inline packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/recipe.py:30]: Does the top-level `pd.Timestamp` typehint not work or `pd.core.api.Timestamp`?

**Review by lewisjared — COMMENTED** [review]
Awesome stuff.

Can you add the tests to the CI. See https://github.com/CMIP-REF/cmip-ref/blob/c3ea68bbc9692056fdde122e872e61041d7da2dd/.github/workflows/ci.yaml#L96

**Comment by lewisjared** [comment]
> @lewisjared Is this the right place to start the ref-metrics-esmvaltool package? Or shall I put it in a separate repository under the [ESMValGroup](https://github.com/esMValGroup/?rgh-link-date=2024-12-13T10%3A19%3A39Z) organization?

This is the right place for now. It can be refactored out in future, but I think it will help the development velocity by building the core and the metric packages in lockstep. It makes refactoring much easier.

> The tests and documentation could better, as could the functionality for editing recipes, running esmvaltool, and the CMEC output bundle, but maybe it's enough at this early stage.

Yeah definitely, but it is good to figure out the sticking points before we dive into adding additional helper functionality.

> Would it make sense to enable building the documentation for pull requests on readthedocs so they are easier to look at? Or are they available from the GitHub Actions and I just missed them?

The CI currently just verifies that the docs build in strict mode. I thought I clicked the button in RtDs to build on each MR, but that might not be true

File: packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/recipe.py (hunk 2 above)
Comment (bouweandela) [reply packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/recipe.py:7]: Thanks for flagging that. I will update it and opened an issue to report this to the authors of pooch: https://github.com/fatiando/pooch/issues/448.

Comment (bouweandela) [reply packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/recipe.py]: Done in 5cdaa67

**Comment by bouweandela** [comment]
> Can you add the tests to the CI. 

Done in c8f31e1. It would be nice to run those using the `make test` command, though I guess that is a bit tricky because of the code coverage collection settings.

File: packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/recipe.py (hunk 3 above)
Comment (bouweandela) [reply packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/recipe.py]: Changed in b3897fd. I'm not very familiar with `pandas`, so I just looked at the runtime type to figure out what it was.

File: packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/example.py (hunk 1 above)
Comment (bouweandela) [reply packages/ref-metrics-esmvaltool/src/ref_metrics_esmvaltool/example.py:6]: It supports roundtripping comments, which seems nice to have.

**Comment by bouweandela** [comment]
> I thought I clicked the button in RtDs to build on each MR, but that might not be true

It seems to work now :partying_face: 

Would it be possible to also enable Codecov comments on pull requests? I can see that it is set up in the README, but it does not tell me which lines have missing coverage.

**Comment by lewisjared** [comment]
> though I guess that is a bit tricky because of the code coverage collection settings.

Correct, otherwise I would have just used `make test`. There is a bit of duplication getting new packages setup

**Comment by lewisjared** [comment]
> Would it be possible to also enable Codecov comments on pull requests? I can see that it is set up in the README, but it does not tell me which lines have missing coverage.

I'll have to look into it because it works for me (although I'm not making PRs from forks). The fork will mean that the `CODECOV_TOKEN` isn't set. I haven't had to support that before so any info would help

**Comment by lewisjared** [comment]
Try merging main as I updated codecov to v5 (#44). That *should* work tokenless

**Comment by bouweandela** [comment]
> Try merging main as I updated codecov to v5 (https://github.com/CMIP-REF/cmip-ref/pull/44). That should work tokenless

It works :partying_face:

**Comment by bouweandela** [comment]
The documentation build failed because downloading the test data was too slow. Downloading several hundred megabytes of data to build the documentation may not be the fastest solution. Maybe we could build the documentation with smaller test data that we put in a GitHub repository?

**Review by lewisjared — APPROVED** [review]

**Pull request merged**